### PR TITLE
Add CRL to http proxy client validation

### PR DESCRIPTION
**Pull request by artificial-aidan**
Signed-off-by: Aidan Jensen <aidan@artificial.com>

Wanted to run this by the maintainers before I put the effort into implementation. Have a possible need to add a certificate revocation list to our http proxies. It gets passed in as a k8s secret just like a CA Certificate, so I would just follow that functionality.

Thoughts or objections?

### CODE REVIEW

**Comment by tsaarni** [comment]
Hi @artificial-aidan 

For me, the idea sounds reasonable.

One thought: In theory CRLs could be large. Using secret as storage could reach [1MB limit](https://kubernetes.io/docs/concepts/configuration/secret/#restrictions). I guess this might happen somewhere around 20000 revoked certs.  Maybe the CRL data could also grow to be substantial part of XDS traffic sent from Contour to Envoys.  Alternative that could scale better, would be to configure file path to CRL but that has its own logistical problems, so secret does feel easier to use.  If the use case is just to revoke few client certs and we can live with that limitation, then I think secret can work.

File: apis/projectcontour/v1/httpproxy.go
Comment (tsaarni) [inline apis/projectcontour/v1/httpproxy.go:932]: I guess CRL file should allow [one or more](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-field-extensions-transport-sockets-tls-v3-certificatevalidationcontext-crl) PEM encoded CRLs concatenated together.

**Comment by artificial-aidan** [comment]
> Hi @artificial-aidan
> 
> For me, the idea sounds reasonable.
> 
> One thought: In theory CRLs could be large. Using secret as storage could reach [1MB limit](https://kubernetes.io/docs/concepts/configuration/secret/#restrictions). I guess this might happen somewhere around 20000 revoked certs. Maybe the CRL data could also grow to be substantial part of XDS traffic sent from Contour to Envoys. Alternative that could scale better, would be to configure file path to CRL but that has its own logistical problems, so secret does feel easier to use. If the use case is just to revoke few client certs and we can live with that limitation, then I think secret can work.

I think CRL's have their own set of problems, and at a large scale like that you are probably better off using a remote list and querying it somehow. If you are managing 20,000 certs as a k8s secret, I think you will have other problems as well :D

**Comment by tsaarni** [comment]
> I think CRL's have their own set of problems, and at a large scale like that you are probably better off using a remote list and querying it somehow. If you are managing 20,000 certs as a k8s secret, I think you will have other problems as well :D

I agree :)

**Comment by sunjayBhatia** [comment]
I would say maybe this should go through a formal design process, just because there are some details here to get right and agree on, but generally yeah agree that is a useful feature

Some details a design could address:
- specifics on how a CRL in a secret is provided (a separate secret as you have here or a field in the existing `caSecret` field for downstream validation)
- how the CRL is validated
- error handling etc. given the documentation on the Envoy field:
```
	// An optional `certificate revocation list
	// <https://en.wikipedia.org/wiki/Certificate_revocation_list>`_
	// (in PEM format). If specified, Envoy will verify that the presented peer
	// certificate has not been revoked by this CRL. If this DataSource contains
	// multiple CRLs, all of them will be used. Note that if a CRL is provided
	// for any certificate authority in a trust chain, a CRL must be provided
	// for all certificate authorities in that chain. Failure to do so will
	// result in verification failure for both revoked and unrevoked certificates
	// from that chain.
	Crl *v3.DataSource `protobuf:"bytes,7,opt,name=crl,proto3" json:"crl,omitempty"`
```

**Comment by artificial-aidan** [comment]
> I would say maybe this should go through a formal design process, just because there are some details here to get right and agree on, but generally yeah agree that is a useful feature
> 
> Some details a design could address:
> 
> * specifics on how a CRL in a secret is provided (a separate secret as you have here or a field in the existing `caSecret` field for downstream validation)
> * how the CRL is validated
> * error handling etc. given the documentation on the Envoy field:
> 
> ```
> 	// An optional `certificate revocation list
> 	// <https://en.wikipedia.org/wiki/Certificate_revocation_list>`_
> 	// (in PEM format). If specified, Envoy will verify that the presented peer
> 	// certificate has not been revoked by this CRL. If this DataSource contains
> 	// multiple CRLs, all of them will be used. Note that if a CRL is provided
> 	// for any certificate authority in a trust chain, a CRL must be provided
> 	// for all certificate authorities in that chain. Failure to do so will
> 	// result in verification failure for both revoked and unrevoked certificates
> 	// from that chain.
> 	Crl *v3.DataSource `protobuf:"bytes,7,opt,name=crl,proto3" json:"crl,omitempty"`
> ```

It would also make sense to include [`only_verify_leaf_cert_crl`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-msg-extensions-transport-sockets-tls-v3-certificatevalidationcontext) in some form. 

I don't know if contour should do CRL validation, does it do CA validation right now? Or does it just pass it to envoy

What is the formal design process? Is that documented somewhere?

**Comment by sunjayBhatia** [comment]
> What is the formal design process? Is that documented somewhere?

https://github.com/projectcontour/community/blob/main/GOVERNANCE.md#proposal-process examples of existing designs in the `design` directory!

**Comment by sunjayBhatia** [comment]
> I don't know if contour should do CRL validation, does it do CA validation right now? Or does it just pass it to envoy

we do some minimal validation to ensure we don't send invalid config to envoy, heres where secret validation is done: https://github.com/projectcontour/contour/blob/main/internal/dag/secret.go

**Comment by artificial-aidan** [comment]
> > I don't know if contour should do CRL validation, does it do CA validation right now? Or does it just pass it to envoy
> 
> we do some minimal validation to ensure we don't send invalid config to envoy, heres where secret validation is done: https://github.com/projectcontour/contour/blob/main/internal/dag/secret.go

Ok, seems reasonable to add a similar validation to CRL bundles, as long as it's easily done through go (assuming it is, haven't tried)

**Comment by tsaarni** [comment]
> Ok, seems reasonable to add a similar validation to CRL bundles, as long as it's easily done through go (assuming it is, haven't tried)

I think for CA bundle we just parse the PEM blocks, not the actual certificates. So I think almost the same code could be used for CRLs as well, but checking for `X509 CRL` PEM block type instead of `CERTIFICATE`. There is also CRL parser in Go, if we would like to go that deep.

**Comment by artificial-aidan** [comment]
Alright, took a stab at a design proposal. First time with this, so let me know if anything obvious is missing.

**Comment by artificial-aidan** [comment]
Bumping this. Any thoughts on the design proposal?

**Review by sunjayBhatia — COMMENTED** [review]
just added some comments around adding some more color to the design, but generally looks great!

File: design/downstream-crl-design.md
Comment (tsaarni) [inline design/downstream-crl-design.md:46]: Could the name be just `crlSecret`? The CRLs are of course published by CAs but I don't recall bumping into CRL being referred as CA CRL.

File: design/downstream-crl-design.md
Comment (tsaarni) [inline design/downstream-crl-design.md:128]: Are you referring to the case which the parameter `only_verify_leaf_cert_crl` addresses (at some level) in the future?  Yes, I agree it is outside of scope for Contour for now - the CRL check is done at level where Contour is not actively involved.

File: design/downstream-crl-design.md
Comment (tsaarni) [inline design/downstream-crl-design.md:101]: I think the type is `X509 CRL` https://datatracker.ietf.org/doc/html/rfc7468#section-4

**Comment by tsaarni** [comment]
I've reviewed the design doc and looks good to me as well!

File: apis/projectcontour/v1/httpproxy.go
Comment (skriss) [inline apis/projectcontour/v1/httpproxy.go:936]: In light of the awkward acronyms, maybe just `crlSecret` (as you have in the design doc) would be better for the YAML, no need to include `ca` as a prefix.

File: design/downstream-crl-design.md
Comment (skriss) [inline design/downstream-crl-design.md:61]: I think @sunjayBhatia mentioned this in one of his comments, but would storing this in the CA cert secret make any sense, or do you prefer to keep it separate? Keeping it separate definitely has the advantage of giving you more space to work with, but just wondering if you'd considered the alternative.

**Review by skriss — COMMENTED** [review]
This proposal makes sense to me and the implementation looks fairly straight-forward. Just had a couple small questions/comments. Thanks @artificial-aidan!

File: apis/projectcontour/v1/httpproxy.go
Comment (artificial-aidan) [reply apis/projectcontour/v1/httpproxy.go:936]: Yep, forgot to update this

File: design/downstream-crl-design.md
Comment (artificial-aidan) [reply design/downstream-crl-design.md:61]: I thought about it, but I think its an added bit of confusion. Having it be a separate secret is clear from a configuration point of view. I guess technically the user could put the same secret for both?

Comment (skriss) [reply design/downstream-crl-design.md:61]: Fair point, as long as the validation allows it which it currently does. In any case I am fine with the separate secret.

**Review by skriss — COMMENTED** [review]
Design doc LGTM - if you drop the API change from this PR, then we can merge just the design, otherwise can wait for the implementation

**Comment by artificial-aidan** [comment]
I'll drop the API change soon. Work got busy, hoping to chip away at it next week.

**Comment by artificial-aidan** [comment]
> Design doc LGTM - if you drop the API change from this PR, then we can merge just the design, otherwise can wait for the implementation

API Change dropped. Ready to merge

**Review by sunjayBhatia — APPROVED** [review]

**Comment by sunjayBhatia** [comment]
hey @artificial-aidan sorry to get back to you late, but this looks good to me 👍🏽 any other @projectcontour/maintainers have comments?

**Review by tsaarni — APPROVED** [review]
Looks good to me too!  

BTW: `only_verify_leaf_cert_crl` option now became available in Envoy 1.21.

**Comment by skriss** [comment]
cc @youngnick planning to merge this tomorrow unless you have any feedback on it.

**Review by youngnick — APPROVED** [review]
LGTM, this is a great design doc, thanks @artificial-aidan!

**Comment by tsaarni** [comment]
Hi @artificial-aidan! I'd be curious to know if you have plans to take a stab at implementing the CRL design proposal? Would be cool to have the feature :)

**Comment by artificial-aidan** [comment]
Priorities shifted at my day job and I have not had a chance to work on it sadly :(

**Comment by tsaarni** [comment]
xref #4252

**Pull request merged**
